funnel/models/membership.py

```diff
@@ -0,0 +1,246 @@
+# -*- coding: utf-8 -*-
+
+from sqlalchemy.ext.hybrid import hybrid_property
+from sqlalchemy.ext.declarative import declared_attr
+from sqlalchemy.ext.associationproxy import association_proxy
+
+from coaster.sqlalchemy import immutable
+
+from . import BaseMixin, UuidMixin, db
+from .user import User
+from .profile import Profile
+from .project import Project
+
+__all__ = ['ProfileAdminMembership', 'ProjectCrewMembership']
+
+
+class ImmutableMembershipMixin(UuidMixin, BaseMixin):
+    """
+    Support class for immutable memberships
+    """
+    __uuid_primary_key__ = True
+    #: List of columns that will be copied into a new row when a membership is amended
+    __role_columns__ = ()
+    #: Parent column ('profile_id' or 'project_id' in the subclasses)
+    __parent_column__ = None
+
+    #: Start time of membership, ordinarily a mirror of created_at except
+    #: for records created when the member table was added to the database
+    granted_at = immutable(db.Column(db.DateTime, nullable=False, default=db.func.utcnow()))
+    #: End time of membership, ordinarily a mirror of updated_at
+    revoked_at = db.Column(db.DateTime, nullable=True)
+
+    @declared_attr
+    def revoked_by_id(cls):
+        """Id of user who revoked the membership"""
+        return db.Column(None, db.ForeignKey('user.id', ondelete='SET NULL'), nullable=True)
+
+    @declared_attr
+    def revoked_by(cls):
+        """User who revoked the membership"""
+        return db.relationship(User, foreign_keys=[cls.revoked_by_id])
+
+    @hybrid_property
+    def active(self):
+        return self.revoked_at is not None
+
+    @active.expression
+    def active(cls):
+        return cls.revoked_at != None  # NOQA
+
+    @declared_attr
+    def __table_args__(cls):
+        return (db.Index(
+            cls.__tablename__ + '_active', cls.__parent_column__, 'user_id',
+            unique=True,
+            postgresql_where=db.text('revoked_at IS NOT NULL')
+        ),)
+
+    # Subclasses must gate these methods in __roles__
+
+    def revoke(self, actor):
+        self.revoked_at = db.func.utcnow()
+        self.revoked_by = actor
+
+    def replace(self, actor, **roles):
+        if not set(roles.keys()).issubset(self.__role_columns__):
+            raise AttributeError("Unknown role")
+        self.revoked_at = db.func.utcnow()
+        self.revoked_by = actor
```

Comment:
This can call the revoke function above?

Reply:
It could.

1. I'm evaluating having an additional column for the revocation reason, to indicate termination or replacement.

2. This code was originally a state manager transition, so one method couldn't call the other. I removed the state manager because it was too verbose for what the model needed.